### DISCUSSION TITLE
xmake v2.9.7

### DISCUSSION THREAD

**Discussion by waruqi** [xmake-io/xmake#5987] [Announcements]
### New features

* [#5813](https://github.com/xmake-io/xmake/pull/5813): Add `before_config` and `after_config` for rule
* [#5848](https://github.com/xmake-io/xmake/issues/5848): Support custom MSVC build tools, e.g. PortableBuildTools and msvc-wine
* [#5880](https://github.com/xmake-io/xmake/pull/5880): Use msvc package to build project
* [#5884](https://github.com/xmake-io/xmake/issues/5884): Add installtips for package
* [#5894](https://github.com/xmake-io/xmake/issues/5894): Add package.merge_staticlibs policy to merge package libs
* [#5948](https://github.com/xmake-io/xmake/pull/5948): Add `lua.native-object` rule
* [#5911](https://github.com/xmake-io/xmake/issues/5911): Support for nuget packages

### Changes

* [#5817](https://github.com/xmake-io/xmake/pull/5817): Improve default pic for packages
* [#5869](https://github.com/xmake-io/xmake/pull/5869): Add libstdc++ standard library modules support for gcc
* [#5923](https://github.com/xmake-io/xmake/pull/5923): Solve the package version and configs conflict

### Bugs fixed

* [#5856](https://github.com/xmake-io/xmake/issues/5856): Fix compile c++modules for clang
* [#5858](https://github.com/xmake-io/xmake/issues/5858): Fix the precompiled header for gcc

### 新特性

* [#5813](https://github.com/xmake-io/xmake/pull/5813): 为 rule 添加 `before_config` 和 `after_config`
* [#5848](https://github.com/xmake-io/xmake/issues/5848): 支持自定义 MSVC 构建工具, PortableBuildTools 和 msvc-wine
* [#5880](https://github.com/xmake-io/xmake/pull/5880): 支持使用 msvc 包去构建工程
* [#5884](https://github.com/xmake-io/xmake/issues/5884): 为包添加自定义安装提示
* [#5894](https://github.com/xmake-io/xmake/issues/5894): 添加 package.merge_staticlibs 策略去合并包安装的静态库
* [#5948](https://github.com/xmake-io/xmake/pull/5948): 添加 `lua.native-object` 规则
* [#5911](https://github.com/xmake-io/xmake/issues/5911): 支持 nuget 包集成

### 改进

* [#5817](https://github.com/xmake-io/xmake/pull/5817): 改进安装包的默认 pic 配置
* [#5869](https://github.com/xmake-io/xmake/pull/5869): 为 gcc 添加 libstdc++ 标准库模块的支持
* [#5923](https://github.com/xmake-io/xmake/pull/5923): 解决包依赖链中版本和配置冲突

### Bugs 修复

* [#5856](https://github.com/xmake-io/xmake/issues/5856): 修复 c++modules 在 clang 下的编译
* [#5858](https://github.com/xmake-io/xmake/issues/5858): 修复 gcc 的头文件预编译问题

## What's Changed
* Add riscv64 support for Android NDK projects by @ifarbod in https://github.com/xmake-io/xmake/pull/5772
* fix error when generating cmake by @cmolocznik in https://github.com/xmake-io/xmake/pull/5779
* fix bindir #5776 by @waruqi in https://github.com/xmake-io/xmake/pull/5780
* fix linker and compiler flags by @waruqi in https://github.com/xmake-io/xmake/pull/5782
* fix: #5777 nvcc warning by @Aurorabili in https://github.com/xmake-io/xmake/pull/5785
* allow using zig cc to compile assembly by @wenxuanjun in https://github.com/xmake-io/xmake/pull/5787
* improve vs/vsxmake group by @waruqi in https://github.com/xmake-io/xmake/pull/5795
* fix package by @waruqi in https://github.com/xmake-io/xmake/pull/5803
* Improve user experience by @Elite-stay in https://github.com/xmake-io/xmake/pull/5804
* add before/after_config #5791 by @waruqi in https://github.com/xmake-io/xmake/pull/5813
* improve default pic for packages by @waruqi in https://github.com/xmake-io/xmake/pull/5817
* add kylin support for linuxos module by @jinke18 in https://github.com/xmake-io/xmake/pull/5834
* improve cmake flags by @waruqi in https://github.com/xmake-io/xmake/pull/5827
* Support custom vc build tools by @star-hengxing in https://github.com/xmake-io/xmake/pull/5823
* add wine run support by @star-hengxing in https://github.com/xmake-io/xmake/pull/5849
* fix `WindowsSDKDir` -> `WindowsSdkDir` by @star-hengxing in https://github.com/xmake-io/xmake/pull/5851
* fix: reset locked repo before pulling by @qudix in https://github.com/xmake-io/xmake/pull/5853
* fix cmake flags by @waruqi in https://github.com/xmake-io/xmake/pull/5850
* feat(msys2): add `xmake.cmd` and `xmake.ps1` for better integration with Windows cmd and powershell by @DeepChirp in https://github.com/xmake-io/xmake/pull/5854
* improve pacman packages by @waruqi in https://github.com/xmake-io/xmake/pull/5855
* fix package linkgroups #5806 by @waruqi in https://github.com/xmake-io/xmake/pull/5866
* maybe a bug: batchcmds by @jinke18 in https://github.com/xmake-io/xmake/pull/5875
* add symlink support for installfiles when using xpack by @jinke18 in https://github.com/xmake-io/xmake/pull/5876
* Fix msvc-wine with cmake package by @star-hengxing in https://github.com/xmake-io/xmake/pull/5871
* Improve copy pdb with cmake & ninja package by @star-hengxing in https://github.com/xmake-io/xmake/pull/5877
* Improve msvc toolchain for package by @waruqi in https://github.com/xmake-io/xmake/pull/5880
* add debugsource param to rule utils.glsl2spv by @dddddjent in https://github.com/xmake-io/xmake/pull/5879
* Add support for libstdc++ standard library modules by @Juvwxyz in https://github.com/xmake-io/xmake/pull/5869
* fix cmake & mingw detect cross-compilation on windows by @star-hengxing in https://github.com/xmake-io/xmake/pull/5888
* fix gcc modules by @waruqi in https://github.com/xmake-io/xmake/pull/5887
* add unknown package tips by @waruqi in https://github.com/xmake-io/xmake/pull/5889
* fix gcc pcheaders by @waruqi in https://github.com/xmake-io/xmake/pull/5896
* improve QML Plugin by @Doekin in https://github.com/xmake-io/xmake/pull/5903
* fix gcc pch #5858 by @waruqi in https://github.com/xmake-io/xmake/pull/5902
* Fix clang module parallel compile error. by @vrecluse in https://github.com/xmake-io/xmake/pull/5899
* Improve to package deps by @waruqi in https://github.com/xmake-io/xmake/pull/5901
* Add Werror and dry-run flags for clang-format by @jj683 in https://github.com/xmake-io/xmake/pull/5908
* Obtain MinGW's path from LLVM_MINGW_DIR/ROOT by @ifarbod in https://github.com/xmake-io/xmake/pull/5909
* Fix meson package build by @star-hengxing in https://github.com/xmake-io/xmake/pull/5912
* Limit adding /subsystem:windows to link.exe by @ifarbod in https://github.com/xmake-io/xmake/pull/5910
* Fix tinycc asm toolset by @star-hengxing in https://github.com/xmake-io/xmake/pull/5914
* Pass --sdk to dependencies by @UE4SS in https://github.com/xmake-io/xmake/pull/5918
* Use ccache when debug cmake package by @star-hengxing in https://github.com/xmake-io/xmake/pull/5919
* Fix /sourceDependencies <path> not on one line when using msvc-wine by @UE4SS in https://github.com/xmake-io/xmake/pull/5920
* Improve zigcc support by @Redbeanw44602 in https://github.com/xmake-io/xmake/pull/5921
* Solve the package configs conflict by @waruqi in https://github.com/xmake-io/xmake/pull/5923
* Merge staticlibs for package by @waruqi in https://github.com/xmake-io/xmake/pull/5927
* fix speed test under non-english linux. by @Redbeanw44602 in https://github.com/xmake-io/xmake/pull/5929
* Fix package.merge_staticlibs by @star-hengxing in https://github.com/xmake-io/xmake/pull/5936
* fix loongarch support under zig toolchain. by @Redbeanw44602 in https://github.com/xmake-io/xmake/pull/5943
* add rule lua-native-object by @Freed-Wu in https://github.com/xmake-io/xmake/pull/5948
* Improve protobuf test by @star-hengxing in https://github.com/xmake-io/xmake/pull/5955
* improve package envs by @waruqi in https://github.com/xmake-io/xmake/pull/5959
* rename luarocks.module to lua.module, platform.linux.driver to platform.linux.module by @Freed-Wu in https://github.com/xmake-io/xmake/pull/5953
* fix nsis #5952 by @waruqi in https://github.com/xmake-io/xmake/pull/5964
* Fix incorrect Zig-CC toolchain wrapper path on Windows by @Doekin in https://github.com/xmake-io/xmake/pull/5967
* improve to parse linux module flags by @waruqi in https://github.com/xmake-io/xmake/pull/5975
* Clang: fix target triple for MinGW by @Doekin in https://github.com/xmake-io/xmake/pull/5980
* Support for nuget packages by @waruqi in https://github.com/xmake-io/xmake/pull/5981
* Fix typos about lua.native-objects by @Freed-Wu in https://github.com/xmake-io/xmake/pull/5985

## New Contributors
* @cmolocznik made their first contribution in https://github.com/xmake-io/xmake/pull/5779
* @Aurorabili made their first contribution in https://github.com/xmake-io/xmake/pull/5785
* @wenxuanjun made their first contribution in https://github.com/xmake-io/xmake/pull/5787
* @qudix made their first contribution in https://github.com/xmake-io/xmake/pull/5853
* @DeepChirp made their first contribution in https://github.com/xmake-io/xmake/pull/5854
* @dddddjent made their first contribution in https://github.com/xmake-io/xmake/pull/5879
* @Juvwxyz made their first contribution in https://github.com/xmake-io/xmake/pull/5869
* @Doekin made their first contribution in https://github.com/xmake-io/xmake/pull/5903
* @vrecluse made their first contribution in https://github.com/xmake-io/xmake/pull/5899
* @jj683 made their first contribution in https://github.com/xmake-io/xmake/pull/5908
* @UE4SS made their first contribution in https://github.com/xmake-io/xmake/pull/5918

**Full Changelog**: https://github.com/xmake-io/xmake/compare/v2.9.6...v2.9.7

<hr /><em>This discussion was created from the release <a href='https://github.com/xmake-io/xmake/releases/tag/v2.9.7'>xmake v2.9.7</a>.</em>

**Comment by plutolove** [comment]
新版本好像有bug，同样的代码编译，依赖了llvm.so，v2.9.6执行没问题，v2.9.7执行报错，提示so找不到，ldd查看链接找不到so，但是编译命令是对的

**Reply by plutolove** [reply]
是不是xmake run自动加了啥环境变量，新版本没了

**Reply by waruqi** [reply]
你拉源码 然后 通过 https://github.com/xmake-io/xmake/pull/5462 这个 定位下，哪行提交导致

**Reply by waruqi** [reply]
另外有问题，提 issues

**Comment by taynpg** [comment]
xp下载哪个，我试了win32可以安装那个不能运行。

**Reply by waruqi** [reply]
默认 win32 安装包就行，按理应该可用

**Reply by taynpg** [reply]
![image](https://github.com/user-attachments/assets/6c9a16f5-146c-4476-9a44-9b252cf620a0)

我是虚拟机跑的，不知道是不是这个问题。

**Reply by star-hengxing** [reply]
这个应该是常见问题，可以考虑引入 yy-thunks https://github.com/Chuyu-Team/YY-Thunks/issues/26

**Reply by waruqi** [reply]
不记得有用过 FlsAlloc，并且 2.7.6 明确支持 xp 可以跑通。。需要排查下期间引入了啥 api 和改动。。去掉就行了

**Reply by waruqi** [reply]
应该是后来 ci 改了，切到 vs2022 编译导致。。得单独搞一路 走 vs2019+v140 编译出包才行。。比较蛋疼

**Reply by star-hengxing** [reply]
考虑引入 vc-ltl5 (减少二进制，链接到 msvcrt) + yy-thunks （补全新 windows api）

**Reply by waruqi** [reply]
那还得引入额外的依赖库，就为了几乎不会用到的 xp，不划算。

我 ci 单独对 2019 编译的包做了上传， xp 上能跑，后面也会上传到 releases 。

https://github.com/xmake-io/xmake/actions/runs/12674416738/artifacts/2402644767

xmake-latest.win32xp.exe
